lib/active_record/connection_adapters/spanner/database_statements.rb

```diff
@@ -41,12 +41,13 @@ def execute sql, name = nil, binds = []
               ["#{v.name}_#{i + 1}", v.type.serialize(v.value)]
             end.to_h
             ActiveSupport::Dependencies.interlock.permit_concurrent_loads do
-              @connection.execute_query(
-                sql,
-                params: params,
-                types: types,
-                transaction_required: transaction_required
-              )
+              if transaction_required
+                transaction do
```

Comment:
Moving this out of the `execute_query` method in the `Connection` class significantly simplifies retrying these kind of statements, as they can rely on the default retry mechanisms built into transactions.